frames/src/main/java/com/checkout/frames/component/cvv/CvvViewModel.kt

```diff
@@ -0,0 +1,126 @@
+import com.checkout.validation.api.CardValidator
+import com.checkout.validation.model.ValidationResult
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+import javax.inject.Provider
+
+internal class CvvViewModel @Inject constructor(
+    private val paymentStateManager: PaymentStateManager,
+    private val cardValidator: CardValidator,
+    private val inputComponentStyleMapper: Mapper<InputComponentStyle, InputComponentViewStyle>,
+    private val inputComponentStateMapper: Mapper<InputComponentStyle, InputComponentState>,
+    style: CvvComponentStyle
+) : ViewModel() {
+
+    val componentState = provideViewState(style)
+    val componentStyle = provideViewStyle(style.inputStyle)
+
+    // Flag to determine that component has been already focused before
+    // Needed to prevent validation on focus switch for initial component state
+    private var wasFocused = false
+
+    internal companion object {
+        val onlyDigitsRegex = "[^0-9]".toRegex()
+    }
+
+    fun prepare() {
+        viewModelScope.launch {
+            paymentStateManager.cardScheme.collect {
+                validateCvv()
+                componentState.cvvLength.value = it.cvvLength.max()
+            }
+        }
+    }
+
+    /**
+     * Make CVV validation, when focus switched to another view.
+     */
+    fun onFocusChanged(isFocused: Boolean) {
+        if (isFocused) wasFocused = isFocused
+
+        if (!isFocused && wasFocused) validateCvv()
+    }
+
+    /**
+     * Update mutable state of input field value.
+     */
+    fun onCvvChange(text: String) = with(text.replace(onlyDigitsRegex, "")) {
+        componentState.cvv.value = this
+        paymentStateManager.cvv.update { this }
+        componentState.hideError()
+    }
+
+    /**
+     * Validate cvv only when it has been already focused before
+     */
+    private fun validateCvv() {
+        if (wasFocused) {
+            val validationResult = cardValidator.validateCvv(
+                componentState.cvv.value,
+                paymentStateManager.cardScheme.value
+            )
+
+            paymentStateManager.isCvvValid.update { validationResult.isValid() }
+            handleValidationResult(validationResult)
+        }
+    }
+
+    private fun handleValidationResult(result: ValidationResult<Unit>) = when (result) {
+        is ValidationResult.Success -> componentState.hideError()
+        is ValidationResult.Failure -> componentState.showError(R.string.cko_cvv_component_error)
+    }
+
+    private fun provideViewState(style: CvvComponentStyle): CvvComponentState =
+        CvvComponentState(inputComponentStateMapper.map(style.inputStyle))
+
+    private fun provideViewStyle(inputStyle: InputComponentStyle): InputComponentViewStyle {
+        var viewStyle = inputComponentStyleMapper.map(inputStyle)
+        val keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number, imeAction = ImeAction.Next)
```

Comment:
Can we add imeaction as ImeAction.Done because the next component would be the billing address which is in form of button action.

Reply:
It's not a solution, because we will have a dynamic component position.
imeAction should be added to the style and changed according to an order which will be provided by the merchant.
It will be done in the scope of prebuild UI.